hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/NodeStakeRepository.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2019-2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.mirror.grpc.repository;
+
+import com.hedera.mirror.common.domain.addressbook.NodeStake;
+import com.hedera.mirror.grpc.config.CacheConfiguration;
+import java.util.Optional;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+
+public interface NodeStakeRepository extends CrudRepository<NodeStake, NodeStake.Id> {
+
+    @Query(value = "select max(consensus_timestamp) from node_stake", nativeQuery = true)
+    Optional<Long> findLatestTimeStamp();
+
+    // All results, including nulls, are cached as not all nodes may be present in the table
+    @Cacheable(cacheManager = CacheConfiguration.NODE_STAKE_CACHE, cacheNames = "node_stake")
+    @Query(value = "select stake from node_stake where consensus_timestamp = ? and node_id = ? ", nativeQuery = true)
+    Optional<Long> findStakeByConsensusTimestampAndNodeId(long consensusTimestamp, long nodeId);
```

Comment:
I don't think we should do N+1 queries, even if cached. It's inefficient and relies upon reflection. Instead, we should do a similar query as the address book entries and get the same N at once. If we move the map construction into the repository it can also be cached.

```suggestion
    Map<Long, Long> findMapByConsensusTimestampAndNodeId(long consensusTimestamp, long nodeId, long limit) {
        return findByConsensusTimestampAndNodeId(consensusTimestamp, nodeId, limit)
            .stream()
            .collect(toMap(NodeStake::getNodeId, NodeStake::getStake);
    }

    @Query(value = "select stake from node_stake where consensus_timestamp = ? and node_id >= ? order by node_id asc limit ?", nativeQuery = true)
    List<NodeStake> findByConsensusTimestampAndNodeId(long consensusTimestamp, long nodeId, long limit);
```

Reply:
Unless every address book entry is guaranteed to have a row in in `node_stake`, the paging won't be in sync. What do you think of [Xin's suggestion](https://github.com/hashgraph/hedera-mirror-node/pull/6249#discussion_r1238782749) on doing this with first and last node ID per page?

Reply:
They will be in sync. If not, there's other serious problems with the network that's out of our control.

Reply:
Okay, I will make it so. The node stake cache still requires its own expiry I think, but perhaps the cache config should reference the existing `cacheSize` property since we're talking the same number of nodes and pages.

Reply:
I believe when a node just gets added to address book, its stake in `node_stake` won't exist until the first UTC midnight afterwards and the behavior also conforms to the stake HIP. The impact is, if we get multiple node stakes, we have to match the node info from address book and the stake from `node_stake` by node id.

Reply:
Interesting. When the node stake update does come, it results in a new `consensus_timestamp` and reflects all of the known nodes in the address book at that time? So, if one was added, or removed, since the previous stake update, then the node stake is now present for all?

Then if a node is added or removed right after, we have to wait for the next stake update to get us back in sync?

Reply:
yes. New rows are added to `node_stake` table when mirrornode processes the `NodeStakeUpdateTransaction`. The transaction is a system transaction generated by consensus nodes daily as the first transaction after each UTC midnight. The values for nodes in the transaction body reflects the staking state at the beginning of that UTC day, or the beginning of the staking period.

Reply:
Is HIP-21 the "staking HIP"?

I think based on the delay between address book updates and node stake updates, it would be possible for the paging to get out of sync. And, once permissionless, community nodes will become a thing. Will they come and go? 

Also, running on other networks, who knows? For example, running locally (`DEMO`), there are four address book entries, but zero rows in `node_stake`, so that would be out of sync. An empty table or missing node ID in `node_stake` is currently handled by assuming node stake of zero.

Reply:
So we can just load all the node stakes into a map instead of paging it. Even if we somehow have 1 million nodes (which we never will since it would be bad for latency and a waste of computing power) that's only 1000000*(8+8)=15MB of memory to store.

Reply:
Do you mean something like [this](https://github.com/hashgraph/hedera-mirror-node/pull/6249/commits/849090318ba2634b719b5ae7646182ac7b52bc4f)?  😄 😄 

Well, maybe I can take another look at `CacheLoader` to find a smoother implementation based on Caffeine...

Reply:
Okay,  my vote is following Xin's suggestion on caching and querying the node stake based on the latest node stake consensus timestamp and the first and last node ID found in the address book entry page. I think this will be efficient and I can code around missing rows in `node_stake`. A missing node ID results in zero stake. Likewise, if the `node_stake` table is empty, then all stake is zero.

Reply:
Not sure why you ever used a scheduled task or a CacheLoader to cache all. You'd simply used a Cacheable on the repo method like you're currently doing. Just need to change the query to find all. I'm also fine with the find using first and last node IDs.

Reply:
I had lookup (by node ID) table on the brain for some reason. Okay, so your suggestion will cache a single `List<NodeStake>`, which is the latest node stake table contents. Then I could iterate through it matching node IDs, rather than do something highly similar a page at a time.

Reply:
Okay, for `NodeStakeRepository`, I have the following:
```
    // An empty result may be cached, indicating the node_stake table is empty
    @Cacheable(cacheManager = CacheConfiguration.NODE_STAKE_CACHE, cacheNames = "node_stake")
    @Query(value = "select * from node_stake where consensus_timestamp = (select max(consensus_timestamp) from "
                    + "node_stake) order by node_id asc",
            nativeQuery = true)
    List<NodeStake> findLatest();
```
This means a couple of things:

1. There is only one cache entry, so I will remove the max cache size property.
2. Basically, the whole table is cached for the expiry period. The find latest timestamp query is part of this query and it won't be issued again until the cache entry expires.
3. Given (2), an expiry of 24 hours could now result in the situation where stale data is kept around for too long after the daily node stake update has occurred.
4. There could be a hybrid implementation, where `findLatestTimestamp()` is invoked on demand (`getNodes()` in the service) and the above query will take a `consensus_timestamp` parameter, which can serve as the cache key. Then there will be no expiry issues.

Reply:
Not sure if I missed part of the discussion here, I think the focus is on how to reduce `N+1` `findMapByConsensusTimestampAndNodeId` calls to just one.

Since we want to get a map instead of a list, this should work

```java
    List<NodeStake> findAllByConsensusTimestamp(long consensusTimestamp);

    @Cacheable(...)
    default Map<Long, Long> findAllStakeByConsensusTimestamp(long consensusTimestamp) {
      return findAllByConsensusTimestamp(consensusTimestamp).stream().collect(Collectors.toMap(NodeStake::getNodeId, NodeStake::getStake));
    }
```

Note you don't need a custom query for `findAllByConsensusTimestamp`.

we still need a max cache size property, perhaps 5, but can't be 1 since concurrent requests to the grpc service around when the node stake is being updated may cause some responses with the stake in the previous staking period and others with the stake in the current period, thus we need at least 2 to avoid evicting the cache back and forth.

Reply:
Oh, I like that! It implies that `findLatestTimestamp()` is still present in the repository and is invoked per `getNodes()` request of the grpc server, making a 24h expiry for the Map appropriate.

Reply:
Done